modules/ingestor/src/service/advisory/csaf/mod.rs

```diff
@@ -1,2 +1,80 @@
+use super::{super::Error, Format};
+use crate::service::advisory::{csaf::loader::CsafLoader, osv::loader::OsvLoader};
+use anyhow::anyhow;
+use bytes::Bytes;
+use futures::{Stream, TryStreamExt};
+use std::time::Instant;
+use trustify_common::db::Transactional;
+use trustify_module_storage::service::{StorageBackend, SyncAdapter};
+
 pub mod loader;
 mod util;
+
+impl super::super::IngestorService {
+    pub async fn ingest<S, E>(&self, source: &str, fmt: Format, stream: S) -> Result<String, Error>
+    where
+        E: std::error::Error,
+        S: Stream<Item = Result<Bytes, E>>,
+    {
+        let start = Instant::now();
+
+        let digest = self
+            .storage
+            .store(stream)
+            .await
+            .map_err(|err| Error::Storage(anyhow!("{err}")))?;
+        let sha256 = hex::encode(digest);
+
+        let storage = SyncAdapter::new(self.storage.clone());
+        let reader = storage
+            .retrieve(sha256.clone())
+            .await
+            .map_err(Error::Storage)?
+            .ok_or_else(|| Error::Storage(anyhow!("file went missing during upload")))?;
+
+        let result = match fmt {
+            Format::CSAF => {
```

Comment:
Not for this PR, but when we add a 3rd format, probably smart to shuffle into a `fn` on `Format` itself to handle the delegation to the appropriate loader?

Reply:
Agree.